### PR TITLE
[dy] Launch with API key

### PR DESCRIPTION
**Pull request by dy46**
# Summary

Launch the data cleaning tool with a group API key to sync pipelines with the Mage database. Calling `mage_ai.launch` with an api_key will kick off the `sync_pipelines()` method which will sync the pipelines between the user's local storage and Mage's database. Right now, we will automatically upload all of the local pipelines to Mage associated to the api key's group.

Usage:
```
import mage_ai

df = pd.read_csv('./testfiles/life_expectancy.csv')

mage_ai.launch(api_key="ErLVkWVZMJVSVDHDpembT7sPpom6cl0Vl8jR1dlg")
mage_ai.clean(df, remote_pipeline_uuid=1)
```

# Tests

tested locally

cc: @wangxiaoyou1993

### CODE REVIEW

File: mage_ai/server/app.py [hunk 1]
```diff
@@ -377,11 +388,18 @@ def launch() -> None:
         'host': host,
         'port': port,
     }
+    app_kwargs = {'port': SERVER_PORT, 'host': 'localhost', 'debug': False}
```
Comment (wangxiaoyou1993) [inline mage_ai/server/app.py:391]: duplicate app_kwargs

File: mage_ai/server/client/mage.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
+import json
+import requests
+from mage_ai.data_cleaner.shared.hash import merge_dict
```
Comment (wangxiaoyou1993) [inline mage_ai/server/client/mage.py:3]: nit: put `from xxx import yyy` above `import xxx`

File: mage_ai/__init__.py [hunk 3]
```diff
@@ -60,11 +61,13 @@ def connect_data(df, name):
     return feature_set
 
 
-def clean(df, pipeline_uuid=None, pipeline_path=None):
+def clean(df, pipeline_uuid=None, pipeline_path=None, remote_pipeline_uuid=None):
```
Comment (wangxiaoyou1993) [inline mage_ai/__init__.py:64]: when calling clean method with remote_pipeline_uuid, we need to pass in api_key as well. it'll be used independently in production environment without launching server.

File: mage_ai/server/app.py [hunk 4]
```diff
@@ -93,7 +94,7 @@ def process():
     if not id:
         return
 
-    feature_set = FeatureSet(id=id)
+    feature_set = FeatureSet(id=id, api_key=api_key)
```
Comment (wangxiaoyou1993) [inline mage_ai/server/app.py:97]: Do we use the process endpoint anywhere?

File: mage_ai/server/client/mage.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                        'X-API-KEY': api_key,
+                    },
+                    url=f'{self.url_prefix}/data_cleaning_pipelines',
+                ).json()
+                pipeline_response = response['data_cleaning_pipeline']
+                pipeline.metadata = merge_dict(pipeline.metadata, {
+                    'remote_id': pipeline_response['id'],
+                })
+        except:
+            pass
```
Comment (wangxiaoyou1993) [inline mage_ai/server/client/mage.py:43]: Let's print out a sync failure message. Same below.

File: mage_ai/server/client/mage.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+    def get_pipeline_actions(self, id, api_key):
```
Comment (wangxiaoyou1993) [inline mage_ai/server/client/mage.py:45]: Same. Check whether api_key is None here.

File: mage_ai/__init__.py (hunk 3 above)
Comment (dy46) [reply mage_ai/__init__.py:64]: I see

File: mage_ai/server/app.py (hunk 4 above)
Comment (dy46) [reply mage_ai/server/app.py:97]: i'm not sure, but i can remove it in a future PR if we don't use this endpoint

**Comment by dy46** [comment]
@wangxiaoyou1993 updated

File: mage_ai/server/app.py [hunk 7]
```diff
@@ -314,16 +319,19 @@ def clean_df(df, name):
...
+def clean_df_with_pipeline(df, id=None, path=None, remote_id=None, mage_api_key=None):
     pipeline = None
     if id is not None:
-        pipeline = Pipeline(id=id)
+        pipeline = Pipeline(id=id).pipeline
     elif path is not None:
-        pipeline = Pipeline(path=path)
+        pipeline = Pipeline(path=path).pipeline
+    elif remote_id is not None:
+        final_key = mage_api_key if mage_api_key is not None else api_key
```
Comment (wangxiaoyou1993) [inline mage_ai/server/app.py:329]: api_key could be undefined.

File: mage_ai/server/data/models.py [hunk 8]
```diff
@@ -226,6 +227,12 @@ def __init__(self, id=None, path=None, feature_set_id=None, pipeline=None):
         if pipeline is not None:
             self.pipeline = pipeline
 
+        if api_key is not None:
```
Comment (wangxiaoyou1993) [inline mage_ai/server/data/models.py:230]: Is this change needed?

Comment (dy46) [reply mage_ai/server/data/models.py:230]: If the api_key is None then I think it makes sense to not try to sync pipelines

Comment (dy46) [reply mage_ai/server/data/models.py:230]: You can initialize the `Pipeline` model class with a `pipeline`, so I guess I was thinking in that case, we would want to sync it. However, I don't think we actually ever initialize a `Pipeline` like that, so we could remove that if it's not necessary.

File: mage_ai/server/app.py (hunk 7 above)
Comment (dy46) [reply mage_ai/server/app.py:328]: @wangxiaoyou1993 good call. I updated to set the `api_key` global variable.

File: mage_ai/server/data/models.py (hunk 8 above)
Comment (wangxiaoyou1993) [reply mage_ai/server/data/models.py:230]: ok, that makes sense. maybe change it to `pipeline is not None and api_key is not None`

**Review by wangxiaoyou1993 — APPROVED** [review]
